### PR TITLE
Add knowledge checks to Rails Advanced Topics

### LINKED ISSUE #22952: Add missing Knowledge Checks, Learning Outcomes, and Knowledge Check links to Ruby on Rails/Advanced Topics
Related to #22470

This task is to go through the Advanced Topics lesson in the Ruby on Rails course and (as applicable):

- Add missing Learning Outcomes
- Add missing Knowledge Checks
- Add answer links in the Knowledge Checks

If the lesson already has Learning Outcomes or Knowledge Checks they do not need to be added to. Many lessons only need answer links added.

Knowledge check links should consist of an `<a>` tag which adds the `knowledge-check-link` class and links to the part of the lesson (or the external resource that is linked in the lesson) where the answer to the Knowledge Check can be found. Knowledge Check links should **not** point to external resources that weren't linked to as part of the lesson content or assignment. You can link to a specific line in the lesson like so:

```html
Lorem ipsum dolor sit amet, <span id='example-knowledge-check'>consectetur adipiscing elit</span>, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua

...
<!-- This a tag just goes to the page where the span with this id lives -->
<a class='knowledge-check-link' href='#example-knowledge-check'>How do you return a value from a function?</a>
```
Or you can link to just a section of the lesson by just linking to the section heading and following [these](https://github.com/TheOdinProject/curriculum/wiki/Contributing-to-Wiki#anchor-linking) rules for determining how to format the link. Example:

```html
### Lorem ipsum dolor sit amet

sed do eiusmod tempor incididunt ut labore et dolore magna aliqua

...
<!-- This a tag just goes to a heading on the page-->
<a class='knowledge-check-link' href='#lorem-ipsum-dolor-sit-amet'>How do you return a value from a function?</a>
```

### Acceptance Criteria

- [ ] Lesson referenced in the title has Learning Outcomes, Knowledge Checks, and Knowledge Check links
- [ ] Knowledge Check links all link to the proper place in the lesson, or the proper external resource that was linked to earlier in the lesson
- [ ] Knowledge Check styling should **not** change

Please do not start work on this unless you've been assigned to it. Comment on this issue and ping me @wise-king-sullyman if you would like to be assigned to it.

Additionally, we would like for contributors to only take on one of these issues at a time. Please only request assignment for (and begin work on) an issue after the Learning Outcome/Knowledge Check/Knowledge Check link issue you're currently assigned to has been approved and merged.

Please check out [this new post on our wiki](https://github.com/TheOdinProject/curriculum/wiki/Learning-Outcomes-vs-Knowledge-Checks) about Learning Outcomes and Knowledge Checks if you need to write new Learning Outcomes or Knowledge Checks for this issue.

### PR DESCRIPTION
**Pull request by msespos**
- [x] You have read [The Odin Projects Contributing Guide](https://github.com/TheOdinProject/curriculum/blob/main/CONTRIBUTING.md).
 - [x] The title summarizes the change and where it happened, for example: "Fixes punctuation in Clean Code lesson".
 - [x] If the PR is related to an open issue, use a [relevant keyword](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) and reference it with the `#` sign and the issue number.
 - [x] If changes are requested, please make the changes in a timely manner. After you submit the changes, request another review from the maintainer (top right).

#### 1. Describe the changes made and include why they are necessary or important:

Add knowledge checks to Rails Advanced Topics

#### 2. Related Issue

Closes #22952

### CODE REVIEW

**Comment by msespos** [comment]
This is my first Knowledge Check edit, so please let me know what needs to be fixed, technically or content-wise. I deliberately used only internal links. Hope that is OK! Thanks.

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
Rather than naming the Knowledge Check ids `knowledge-check-x`  I'd like to see the ids be semantically meaningful.

Additionally, you don't need to use ids for each jumplink and can instead use the associated section anchor link when the key info is located near(ish) the start of the section or the section isn't super long.

I will have to review this lesson before I can do a Knowledge Check/content fit evaluation to finish the review as I haven't done this lesson yet myself😅, so it will probably take me a few days.

If any other maintainers who have already done this lesson see this comment please feel free to finish the review if you have a chance before I'm able to.

**Comment by msespos** [comment]
I think this edit should do it. All links now use section anchor links instead of IDs and the spans have been removed. Let me know if you need further changes. Thanks!

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
One quick note just for the future: please link PRs to issues they will close when applicable, I've already done that here so there's nothing else you need to do, just a note for the future.

Overall I really like the look of these and only have one KC I'd like to see changed.

If you'd be open to doing it I think that the Learning Outcomes could also use a sprucing up as they don't align with our current LO standards, if you'd like to do that you can do it under this PR or make another if you'd like. Of course, you don't have to do it if you don't want to though.

**Comment by msespos** [comment]
> 

Sorry about forgetting the issue link. Thanks for adding it and for the reminder!

I edited the KCs as you requested and also edited the LOs down from 19 (!) to 6. Those were done in two separate commits.

Let me know if you need more done on this. Thanks Sully!

File: rails_programming/mailers_advanced_topics/advanced_topics.md [hunk 1]
```diff
@@ -5,25 +5,12 @@ There are some topics that we just haven't had a chance to get into yet but will
...
-* How do you set up a redirect route that also passes along any parameters?
-* How do you name a route using an alias?
-* Why might you want to use nested or multiple layouts?
-* How would you (roughly) go about implementing this?
-* How can you pass variables between your layouts?
-* How do you `#yield` to `#content_for` content?
+* When do you use a singular Resource in your router and when do you use plural Resources?
+* When would you use nested routes?
+* When would you use a "member" route? A "collection" route?
+* When would you use multiple layouts?
```
Comment (wise-king-sullyman) [inline rails_programming/mailers_advanced_topics/advanced_topics.md:11]: These still aren't *quite* lining up with our current Learning Outcome direction, I would check out [our wiki post about Learning Outcomes and Knowledge Checks](https://github.com/TheOdinProject/curriculum/wiki/Learning-Outcomes-vs-Knowledge-Checks) if you haven't yet.

If you already have but aren't sure what it is we're looking for (it's difficult to describe) I could provide some suggestions, but I don't want to take that agency away from you unless you request it 🙂

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
Looking solid! Just need a little more refinement on the Learning Outcomes.

File: rails_programming/mailers_advanced_topics/advanced_topics.md (hunk 1 above)
Comment (msespos) [reply rails_programming/mailers_advanced_topics/advanced_topics.md:11]: So are you saying that those aren't appropriately high-level questions?

Comment (wise-king-sullyman) [reply rails_programming/mailers_advanced_topics/advanced_topics.md:11]: Correct, that is what I'm saying. As an example of what are think are high level enough "What is metaprogramming?" and "What are Design Patterns?" seem great. If that still isn't clear though please let me know.

**Comment by msespos** [comment]
Hope this one works out. I just generalized the questions to ask what each topic is rather than to ask about applying the knowledge. Let me know - thanks!

**Review by wise-king-sullyman — APPROVED** [review]
These look good to me. Thanks Mike!

**Pull request merged**